datashader/reductions.py

```diff
@@ -43,39 +43,158 @@ def apply(self, df):
+        """Validates input shape"""
+        raise NotImplementedError("validate not implemented")
 
-class category_codes(Preprocess):
+    def apply(self, df):
+        """Applies preprocessor to DataFrame and returns array"""
+        raise NotImplementedError("apply not implemented")
+
+class category_codes(CategoryPreprocess):
     """Extract just the category codes from a categorical column."""
+    def categories(self, input_dshape):
+        return input_dshape.measure[self.column].categories
+
+    def validate(self, in_dshape):
+        if not self.column in in_dshape.dict:
+            raise ValueError("specified column not found")
+        if not isinstance(in_dshape.measure[self.column], ct.Categorical):
+            raise ValueError("input must be categorical")
+
     def apply(self, df):
         if cudf and isinstance(df, cudf.DataFrame):
             return df[self.column].cat.codes.to_gpu_array()
         else:
             return df[self.column].cat.codes.values
 
-class category_values(Preprocess):
-    """Extract multiple columns from a dataframe as a numpy array of values."""
-    def __init__(self, columns):
-        self.columns = list(columns)
+class category_modulo(category_codes):
+    """
+    A variation on category_codes that assigns categories using an integer column, modulo a base.
+    Category is computed as (column_value - offset)%modulo.
+    """
+
+    # couldn't find anything in the datashape docs about how to check if a CType is an integer, so just define a big set
+    IntegerTypes = {ct.bool_, ct.uint8, ct.uint16, ct.uint32, ct.uint64, ct.int8, ct.int16, ct.int32, ct.int64}
+
+    def __init__(self, column, modulo, offset=0):
+        super().__init__(column)
+        self.offset = offset
+        self.modulo = modulo
+
+    def _hashable_inputs(self):
+        return super()._hashable_inputs() + (self.offset, self.modulo)
+
+    def categories(self, in_dshape):
+        return list(range(self.modulo))
+
+    def validate(self, in_dshape):
+        if not self.column in in_dshape.dict:
+            raise ValueError("specified column not found")
+        if in_dshape.measure[self.column] not in self.IntegerTypes:
+            raise ValueError("input must be an integer column")
+
+    def apply(self, df):
+        result = (df[self.column] - self.offset) % self.modulo
+        if cudf and isinstance(df, cudf.DataFrame):
+            return result.to_gpu_array()
+        else:
+            return result.values
+
+class category_binning(category_modulo):
+    """
+    A variation on category_codes that assigns categories by binning a continuously-valued column.
+    The number of categories returned is always nbins+1.
+    The last category (nbin) is for NaNs in the data column, as well as for values under/over the binned
+    interval (when include_under or include_over is False).
+
+    Parameters
+    ----------
+    column:   column to use
+    bin0:     lower bound of first bin
+    binsize:  bin size
```

Comment:
Is it important that the bin size be exactly as specified? I would think that a more natural interface would be to specify the lower and upper bounds of the range and then nbins, with binsize being computed as (upper-lower)/nbins.  That way people can calculate lower and upper from their data and easily try different nbins, whereas how they have to write an expression for binsize to make it work like that.

Reply:
Good suggestion.